hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NSSummaryEndpoint.java

```diff
@@ -0,0 +1,764 @@
+      quotaUsageResponse.setResponseCode(ResponseStatus.PATH_NOT_FOUND);
+    } else { // directory and key are not applicable for this request
+      quotaUsageResponse.setResponseCode(
+              ResponseStatus.TYPE_NOT_APPLICABLE);
+    }
+    return Response.ok(quotaUsageResponse).build();
+  }
+
+  /**
+   * Endpoint that returns aggregate file size distribution under a path.
+   * @param path request path
+   * @return File size distribution response
+   * @throws IOException
+   */
+  @GET
+  @Path("/dist")
+  public Response getFileSizeDistribution(@QueryParam("path") String path)
+          throws IOException {
+
+    if (path == null || path.length() == 0) {
+      return Response.status(Response.Status.BAD_REQUEST).build();
+    }
+
+    String normalizedPath = OmUtils.normalizeKey(path, false);
+    String[] names = parseRequestPath(normalizedPath);
+    EntityType type = getEntityType(path, names);
+
+    FileSizeDistributionResponse distReponse =
+            new FileSizeDistributionResponse();
+    switch (type) {
+    case ROOT:
+      List<OmBucketInfo> allBuckets = listBucketsUnderVolume(null);
+      int[] fileSizeDist = new int[ReconConstants.NUM_OF_BINS];
+
+      // accumulate file size distribution arrays from all buckets
+      for (OmBucketInfo bucket : allBuckets) {
+        long bucketObjectId = bucket.getObjectID();
+        int[] bucketFileSizeDist = getTotalFileSizeDist(bucketObjectId);
+        // add on each bin
+        for (int i = 0; i < ReconConstants.NUM_OF_BINS; ++i) {
+          fileSizeDist[i] += bucketFileSizeDist[i];
+        }
+      }
+      distReponse.setFileSizeDist(fileSizeDist);
+      break;
+    case VOLUME:
+      List<OmBucketInfo> buckets = listBucketsUnderVolume(names[0]);
+      int[] volumeFileSizeDist = new int[ReconConstants.NUM_OF_BINS];
+
+      // accumulate file size distribution arrays from all buckets under volume
+      for (OmBucketInfo bucket : buckets) {
+        long bucketObjectId = bucket.getObjectID();
+        int[] bucketFileSizeDist = getTotalFileSizeDist(bucketObjectId);
+        // add on each bin
+        for (int i = 0; i < ReconConstants.NUM_OF_BINS; ++i) {
+          volumeFileSizeDist[i] += bucketFileSizeDist[i];
+        }
+      }
+      distReponse.setFileSizeDist(volumeFileSizeDist);
+      break;
+    case BUCKET:
+      long bucketObjectId = getBucketObjectId(names);
+      int[] bucketFileSizeDist = getTotalFileSizeDist(bucketObjectId);
+      distReponse.setFileSizeDist(bucketFileSizeDist);
+      break;
+    case DIRECTORY:
+      long dirObjectId = getDirObjectId(names);
+      int[] dirFileSizeDist = getTotalFileSizeDist(dirObjectId);
+      distReponse.setFileSizeDist(dirFileSizeDist);
+      break;
+    case KEY:
+      // key itself doesn't have file size distribution
+      distReponse.setStatus(ResponseStatus.TYPE_NOT_APPLICABLE);
+      break;
+    case UNKNOWN:
+      distReponse.setStatus(ResponseStatus.PATH_NOT_FOUND);
```

Comment:
I think you still need to setFileSizeDist, probably to an empty array for UNKNOWN and KEY cases. Otherwise you get NPE:

![image](https://user-images.githubusercontent.com/50227127/127206894-4b856184-3ef1-4bec-98bb-208dd02fc391.png)